docs/rfcs/20220829-distributed-dynamic-routing.md

```diff
@@ -0,0 +1,213 @@
+   ┌──────┐        ┌──────┐         ┌──────┐    │       │    ┌──────┐        ┌──────┐         ┌──────┐   
+┌──┤Node0 ├──┐  ┌──┤Node1 ├──┐   ┌──┤Node2 ├──┐ │ Shard │ ┌──┤Node0 ├──┐  ┌──┤Node1 ├──┐   ┌──┤Node2 ├──┐
+│  └──────┘  │  │  └──────┘  │   │  └──────┘  │ │ Load  │ │  └──────┘  │  │  └──────┘  │   │  └──────┘  │
+│┌─────────┐ │  │┌─────────┐ │   │┌─────────┐ │ │Balance│ │┌─────────┐ │  │┌─────────┐ │   │┌─────────┐ │
+││Shard0 L │ │  ││Shard0 F │ │   ││Shard1 L │ │ │       │ ││Shard0 L │ │  ││Shard0 F │ │   ││Shard1 L │ │
+│└─────────┘ │  │└─────────┘ │   │└─────────┘ │ └───────┘ │└─────────┘ │  │└─────────┘ │   │└─────────┘ │
+│┌─────────┐ │  │┌─────────┐ │   │            │──────────▶│┌─────────┐ │  │┌─────────┐ │   │┌─────────┐ │
+││Shard2 L │ │  ││Shard1 F │ │   │            │           ││Shard2 L │ │  ││Shard1 F │ │   ││Shard2 F │ │
+│└─────────┘ │  │└─────────┘ │   │            │           │└─────────┘ │  │└─────────┘ │   │└─────────┘ │
+│            │  │┌─────────┐ │   │            │           │            │  │            │   │            │
+│            │  ││Shard2 F │ │   │            │           │            │  │            │   │            │
+│            │  │└─────────┘ │   │            │           │            │  │            │   │            │
+└────────────┘  └────────────┘   └────────────┘           └────────────┘  └────────────┘   └────────────┘
+```
+
+
+
+### Shard Change Process
+In CeresDB, we could change `Replication_Factor` to determine the number of shard replication. Therefore, in the processing of shard, we also need to consider the number of shard replication
+
+
+
+#### Switch Leader
+
+Only Leader Shard(`Replication_Factor` == 1)
+
+1. When there is only one leader shard, there is no need to switch leader, CeresMeta only need to migrate shard to another node.
+
+
+Multi Follower Shard(`Replication_Factor` >= 2)
+
+1. First, set the original leader shard and the follwer to be switched to PENDING_ FOLLOWER and PENDING_ LEADER.
+1. Wait for the leader shard to complete the preparatory action before the leader switch, and notify the CeresMeta after the completion. In this process, the new leader cannot process the write request.
+1. After receiving this request, CeresMeta will update the status of the two shards, and the leader switch is completed.
+
+
+#### Shard Migrate
+
+Only Leader Shard(`Replication_Factor` == 1)
+1. Same as `Multi Follower Shard` leader shard.
+
+
+Multi Follower Shard(`Replication_Factor` >= 2)
+- Leader Shard：
+    1. First, CeresMeta adjusts the routing relationship and create a new follower shard on the node.
+    1. Follower shard notifies CeresMeta after initialization.
+    1. CeresMeta started the process of leader switch, switched the leader to the newly created follower shard.
+    1. Close the original leader shard after the leader switch is completed.
+- Follower Shard：
+    1. Adjust shard node mapping to directly assign this shard to another node.
```

Comment:
Maybe expand an tmp Follower shard to ensure fllower read capability. After shard migration, shrink the tmp Follower shard.

Reply:
> Maybe expand an tmp Follower shard to ensure fllower read capability. After shard migration, shrink the tmp Follower shard.

Yes, I didn't describe the use of this tmp follower clearly, I will add more clearly description